File: Python/Demos/demos.py
import sys
import os

# Get the parent directory and add it to sys.path
current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.dirname(current_dir)
sys.path.append(parent_dir)

from Library.weather import Weather


weather = Weather(api_key='3KAJKHWT3UEMRQWF2ABKVVVZE')
    
# fetch from the Visual Crossing API
weather.fetch_weather_data("38.95,-95.664", "2023-01-01", '2023-01-02', include='hours')


# using the method [get_weather_data] of the class to get the weather data
print(">>Result: ", weather.get_weather_data(['timezone']))

# using the method [get_tzoffset] and [set_tzoffset] of the class to set the `tzoffset`
weather.set_tzoffset(-3.0)
print(">>changed tzoffset: ", weather.get_tzoffset())

# using the method [get_weather_hourly_data] of the class to get the daily  data
print(">>Daily Data:", weather.get_weather_daily_data())
print(">>Daily Data with some elements:", weather.get_weather_daily_data(['temp', 'precip']))

# using the method [get_weather_hourly_data] of the class to get the hourly data
# print(">>Hourly Data:", weather.get_weather_hourly_data())
print(">>Hourly Data with some elements:", weather.get_weather_hourly_data(['temp', 'precip']))

# using the method [get_daily_datetimes] of the class to get datetimes of the daily data
print(">>Daily datetimes: ", weather.get_daily_datetimes())

# using the method [get_hourly_datetimes] of the class to get datetimes of the daily data
# print(">>Hourly datetimes: ", weather.get_hourly_datetimes())

# using the method [get_data_on_day] of the class to get data of the specific day within `days` key
print(">>Weather data with some elements on a day: ", weather.get_data_on_day('2023-01-02', ['temp', 'humidity', 'dew']))

# using the method [get_hourlyData_on_day] of the class to get hourly data of the specific day within `days` key
print(">>Hourly data with some elements on a day: ", weather.get_hourlyData_on_day('2023-01-02', ['temp', 'humidity']))

# using the method [get_tempmax_on_day] of the class to get tempmax of the specific day within `days` key
# print(">>tempmax on a day: ", weather.get_tempmax_on_day(1))
# print(">>tempmax on a day: ", weather.get_tempmax_on_day('2023-01-02'))

# using the method [set_tempmax_on_day] of the class to get tempmax of the specific day within `days` key
weather.set_tempmax_on_day('2023-01-02', 18)
print(">>changed tempmax on a day: ", weather.get_tempmax_on_day(1))

# using the method [get_snowdepth_on_day] and [set_snowdepth_on_day] of the class to get and set snowdepth of the specific day within `days` key for daily data
print(">>snowdepth on a day: ", weather.get_snowdepth_on_day(1))
weather.set_snowdepth_on_day('2023-01-02', 1)
print(">>changed snowdepth on a day: ", weather.get_snowdepth_on_day(1))


# using the method [get_data_at_datetime] and [set_data_at_datetime] of the class to get and set data of the specific date and time within `hours` key for hourly data
print(">>Weather data with some elements at a datetime: ", weather.get_data_at_datetime(1,1, ['temp', 'precip', 'humidity']))
weather.set_data_at_datetime('2023-01-02', '01:00:00', {'temp': 25.0})
print(">>Changed weather data at a datetime: ", weather.get_data_at_datetime('2023-01-02', '01:00:00'))

# using the method [get_datetimeEpoch_at_datetime] and [set_datetimeEpoch_at_datetime] of the class to get and set datetimeEpoch of the specific date and time within `hours` key for hourly data
print(">>datetimeEpoch at a datetime: ", weather.get_datetimeEpoch_at_datetime(1,1))
weather.set_datetimeEpoch_at_datetime('2023-01-02', 1, '1111111111')
print(">>changed datetimeEpoch at a datetime: ", weather.get_datetimeEpoch_at_datetime('2023-01-02', '01:00:00'))

# using the method [get_stations_at_datetime] of the class to get and set datetimeEpoch of the specific date and time within `hours` key for hourly data
print(">>stations at a datetime: ", weather.get_stations_at_datetime(1,1))

File: MATLAB/Library/demos.m
clear; clc
% Create an instance of the Weather class
w = Weather('Your API Key');

% Set daily weather data manually
daily_data = {struct('temp', 70, 'humidity', 50, 'hours', {struct('temp', 60, 'humidity', 40), struct('temp', 65, 'humidity', 45)}), struct('temp', 72, 'humidity', 55, 'hours', {struct('temp', 66, 'humidity', 43), struct('temp', 68, 'humidity', 48)})};
% w.set_weather_daily_data(daily_data);

% Fetch weather data for a specific location
data = w.fetch_weather_data('New York', '2024-05-01', '2024-05-15', 'us', 'hours');
fprintf('Fetched data: \n');
disp(data)

% Get stored weather data
stored_data = w.get_weather_data();
fprintf('Stored data: \n');
disp(stored_data)

% Get daily weather data, filtered by specific elements
daily_data_filtered = w.get_weather_daily_data({'temp'});
fprintf('Filtered daily data: \n');
dispCellArrayOfStructs(daily_data_filtered)

% Get hourly weather data, filtered by specific elements
hourly_data_filtered = w.get_weather_hourly_data({'temp', 'humidity'});
fprintf('Filtered hourly data: \n');
dispCellArrayOfStructs(hourly_data_filtered);

% Set and get query cost manually
w.set_queryCost(5.0);
query_cost = w.get_queryCost();
disp(['Query cost: ', num2str(query_cost)]);

% Set and get latitude
w.set_latitude(40.7128);
latitude = w.get_latitude();
disp(['Latitude: ', num2str(latitude)]);

% Set and get longitude
w.set_longitude(-74.0060);
longitude = w.get_longitude();
disp(['Longitude: ', num2str(longitude)]);

% Set and get resolved address
w.set_address('New York, NY, USA');
resolved_address = w.get_resolvedAddress();
disp(['Resolved address: ', resolved_address]);

% Set and get timezone
w.set_timezone('EST');
timezone = w.get_timezone();
disp(['Timezone: ', timezone]);

% Set and get timezone offset
w.set_tzoffset(-5.0);
tzoffset = w.get_tzoffset();
disp(['Timezone Offset: ', num2str(tzoffset)]);

% Set and get weather stations
w.set_stations({'Station1', 'Station2', 'Station3'});
stations = w.get_stations();
disp('Weather Stations:');
disp(stations);

% Get daily datetimes
daily_datetimes = w.get_daily_datetimes();
disp('Daily Datetimes:');
disp_array_summary(daily_datetimes);

% Get hourly datetimes
hourly_datetimes = w.get_hourly_datetimes();
disp('Hourly Datetimes:');
disp_array_summary(hourly_datetimes);

% Get and set snow depth on a specific day
snowdepth = w.get_snowdepth_on_day('2024-05-05');
disp(['Snow Depth: ', num2str(snowdepth)]);
w.set_snowdepth_on_day('2024-05-05', 12.5);
disp(['Changed snow Depth: ', num2str(w.get_snowdepth_on_day('2024-05-05'))]);

% Get and set wind gust
windgust = w.get_windgust_on_day('2024-05-05');
disp(['Wind Gust: ', num2str(windgust)]);
w.set_windgust_on_day('2024-05-05', 25.0);

% Get and set wind speed on a specific day
windspeed = w.get_windspeed_on_day('2024-05-05');
disp(['Wind Speed: ', num2str(windspeed)]);
w.set_windspeed_on_day('2024-05-05', 15.0);

% Get and set wind direction on a specific day
winddir = w.get_winddir_on_day('2024-05-05');
disp(['Wind Direction: ', num2str(winddir)]);
w.set_winddir_on_day('2024-05-05', 180);

% Get and set pressure on a specific day
pressure = w.get_pressure_on_day('2024-05-05');
disp(['Pressure: ', num2str(pressure)]);
w.set_pressure_on_day('2024-05-05', 1015);

% Get and set cloud cover on a specific day
cloudcover = w.get_cloudcover_on_day('2024-05-05');
disp(['Cloud Cover: ', num2str(cloudcover)]);
w.set_cloudcover_on_day('2024-05-05', 80);

% Get and set visibility on a specific day
visibility = w.get_visibility_on_day('2024-05-05');
disp(['Visibility: ', num2str(visibility)]);
w.set_visibility_on_day('2024-05-05', 10);

% Get and set solar radiation on a specific day
solarradiation = w.get_solarradiation_on_day('2024-05-05');
disp(['Solar Radiation: ', num2str(solarradiation)]);
w.set_solarradiation_on_day('2024-05-05', 900);

% Get and set data at a specific datetime
dt_data = w.get_data_at_datetime('2024-05-01', '00:00:00', {'datetime', 'temp', 'dew'});
fprintf('Data at a specific datetime: \n');
disp(dt_data)
w.set_data_at_datetime('2024-05-01', '00:00:00', struct('temp', 24, 'humidity', 50));
fprintf('Changed data at a specific datetime: \n');
disp(w.get_data_at_datetime('2024-05-01', '00:00:00'))

% Get and set datetimeEpoch at a specific datetime
disp(['Datetime Epoch at as specific datetime: ',num2str(w.get_datetimeEpoch_at_datetime('2024-05-01', '01:00:00'))])

% Get and set datetimeEpoch at a specific datetime
w.set_windspeed_at_datetime('2024-05-01', 2, 5.1)
disp(['Wind speed at as specific datetime: ',num2str(w.get_windspeed_at_datetime('2024-05-01', '01:00:00'))])